enterprise_access/apps/subsidy_access_policy/models.py

```diff
@@ -1228,17 +1234,84 @@ def remaining_balance_per_user(self, lms_user_id):
+
+        return new_policy
+
     def assignment_request_can_redeem(self, lms_user_id, content_key, skip_customer_user_check=False,
                                       skip_enrollment_deadline_check=False, **kwargs
                                       ):
-        policy_instance = AssignedLearnerCreditAccessPolicy()
+        policy_instance = self.copy_context_to(AssignedLearnerCreditAccessPolicy)
         return policy_instance.can_redeem(lms_user_id, content_key, skip_customer_user_check,
                                           skip_enrollment_deadline_check, **kwargs
                                           )
 
     def assignment_request_redeem(self, lms_user_id, content_key, all_transactions, metadata=None, **kwargs):
-        policy_instance = AssignedLearnerCreditAccessPolicy()
-        return policy_instance.redeem(lms_user_id, content_key, all_transactions, metadata=metadata, **kwargs)
+
+        learner_credit_request = kwargs.get('learner_credit_request')
+        logger.info(
+            "LearnerCreditRequestActions redeem record creation requested."
+            "lms_user_id=%s, content_key=%s",
+            lms_user_id,
+            content_key
+        )
+        if self.bnr_enabled and learner_credit_request:
+            logger.info(
+                "Creating LearnerCreditRequestActions record for redemption attempt. "
+                "learner_credit_request_uuid=%s, lms_user_id=%s, content_key=%s",
+                learner_credit_request.uuid,
+                lms_user_id,
+                content_key
+            )
+            action = LearnerCreditRequestActions.create_action(
+                learner_credit_request=learner_credit_request,
+                recent_action=SubsidyRequestStates.ACCEPTED,
+                status=LearnerCreditRequestUserMessages.ACCEPTED,
+            )
+            try:
+                policy_instance = self.copy_context_to(AssignedLearnerCreditAccessPolicy)
+                result = policy_instance.redeem(lms_user_id, content_key, all_transactions, metadata=metadata, **kwargs)
+                logger.info(
+                    "Successfully redeemed content through assignment_request_redeem. "
+                    "lms_user_id=%s, content_key=%s, transaction_uuid=%s",
+                    lms_user_id,
+                    content_key,
+                    result.get('uuid', 'unknown')
+                )
+                return result
+            except Exception as exc:
+                logger.exception(
+                    "Error redeeming content through assignment_request_redeem. "
+                    "learner_credit_request_uuid=%s, lms_user_id=%s, content_key=%s",
+                    learner_credit_request.uuid if learner_credit_request else None,
+                    lms_user_id,
+                    content_key
+                )
+                action.error_reason = LearnerCreditRequestActionErrorReasons.FAILED_REDEMPTION
```

Comment:
We will also need to set the `status` back to the previous state i.e, `APPROVED` in case of errors.